### PR TITLE
pov-recovery: try connect to validators

### PR DESCRIPTION
**Pull request by ordian**
A companion to https://github.com/paritytech/polkadot/pull/6081.

- [x] Make it compile due to unrelated unmerged companions
- [x] Test it works with the polkadot changes
- [x] Revert the REVERTME commits
- [x] Test it doesn't work on master

### CODE REVIEW

File: zombienet_tests/0002-pov_recovery.toml [hunk 1]
```diff
@@ -33,12 +33,12 @@ cumulus_based = true
...
+  args = ["--use-null-consensus", "--disable-block-announcements"]
 
   # run eve as parachain full node that is only connected to dave
   [[parachains.collators]]
   name = "eve"
   validator = false
   image = "{{COL_IMAGE}}"
   command = "test-parachain"
-  args = ["--reserved-only", "--reserved-nodes {{'charlie'|zombie('multiAddress')}}", "--use-null-consensus", "--disable-block-announcements"]
+  args = ["--reserved-only", "--reserved-nodes {{'dave'|zombie('multiAddress')}}",  "--use-null-consensus", "--disable-block-announcements"]
```
Comment (bkchr) [inline zombienet_tests/0002-pov_recovery.toml:44]: We need to pass the `--reserved-node]` flag to the polkadot node, so after the `--`.

Comment (ordian) [reply zombienet_tests/0002-pov_recovery.toml]: How did it work before? :)

**Comment by ordian** [comment]
Not quite sure why the check-rustdoc and runtime-benchmarks fail. Is it related to https://github.com/paritytech/polkadot/pull/5205?

File: zombienet_tests/0002-pov_recovery.toml (hunk 1 above)
Comment (bkchr) [reply zombienet_tests/0002-pov_recovery.toml]: Is zombienet probably adding other nodes as bootnodes automatically? IDK :P

**Comment by bkchr** [comment]
@ordian is there any update? Could you already make it fail without your patch?

**Comment by ordian** [comment]
> @ordian is there any update? Could you already make it fail without your patch?

I'm trying to make it work with my patch first. Investing why it doesn't.

**Comment by skunert** [comment]
@bkchr  I investigated this for a while. Strangely, the validator nodes seem to be connected initially, allowing the collators to recover some blocks. I tried to prevent this by several methods, however none of them worked.

However, I noticed that the connections to the validators time out after a while and reconnection does not happen. So after the initial few blocks, pov-recovery does not work anymore. I ran this test many times locally, and collators were never able to recover more than 14 blocks. So setting the block target to 20 in this test makes it fail consistently.

By changing to `IfDisonnected::Apply` the test consistently succeeds.

There is one more thing: This test shows flaky behaviour in certain situations.
In these zombienet tests, the nodes do not stasrt simultaneously. It can happen that a node starts and misses the relay chain block notification that contains the first parachain candidate. In this situation, the node will be unable to recover all following blocks because the parent is missing. We now register the parachain after 30 relay chain blocks, so everyone should be able to get the relevant notifications.

File: zombienet_tests/register-para.js [hunk 2]
```diff
@@ -0,0 +1,10 @@
+async function run(nodeName, networkInfo, args) {
+    const paraIdStr = args[0];
+    const para = networkInfo.paras[paraIdStr];
+    const relayNode = networkInfo.relay[0];
+
+    console.log(para);
+    await zombie.registerParachain(parseInt(paraIdStr,10),para.wasmPath, para.statePath, relayNode.wsUri, "//Alice", true);
```
Comment (pepoviola) [inline zombienet_tests/register-para.js:7]: @ordian / @skunert this needs the last version of zombienet. I will change it in this branch.
Thanks!

**Comment by bkchr** [comment]
I think one more thing that should be looked into is this: https://github.com/paritytech/substrate/blob/0ee03277c33b6334ddba7434e014fa637dcb6107/client/cli/src/params/network_params.rs#L236

I suspect that Kademlia is doing the random walks, aka connecting to nodes "on its own". If we disable this, I assume the node should not to any other node, besides the bootnodes/reserved nodes. Maybe @tomaka / @mxinden could shed some light onto this. TLDR is that we don't want that certain nodes connect to other nodes in any "automatic" way and only when some behavior triggers these connection. Currently it looks like as we are always having a connection between all nodes at some point (I get that this is related to the small size of the test network, however it would be nice to not require to run a big network to test the stuff we want to test :P).

**Comment by ordian** [comment]
> I think one more thing that should be looked into is this: https://github.com/paritytech/substrate/blob/0ee03277c33b6334ddba7434e014fa637dcb6107/client/cli/src/params/network_params.rs#L236

Relay part of collators and all validators are using `--reserved-only`, so it should be disabled.

**Comment by skunert** [comment]
> I think one more thing that should be looked into is this: https://github.com/paritytech/substrate/blob/0ee03277c33b6334ddba7434e014fa637dcb6107/client/cli/src/params/network_params.rs#L236
> 
> I suspect that Kademlia is doing the random walks, aka connecting to nodes "on its own". If we disable this, I assume the node should not to any other node, besides the bootnodes/reserved nodes. Maybe @tomaka / @mxinden could shed some light onto this. TLDR is that we don't want that certain nodes connect to other nodes in any "automatic" way and only when some behavior triggers these connection. Currently it looks like as we are always having a connection between all nodes at some point (I get that this is related to the small size of the test network, however it would be nice to not require to run a big network to test the stuff we want to test :P).

I was suspecting this too and had dht random walks already disabled for my tests. I also modified the discovery limit to 0 [here](https://github.com/paritytech/substrate/blob/6f453b5c2b6d7479189da06fd03e403b95503bdd/client/network/src/service.rs#L275).

When disabled, I do not see this message anymore, as expected.
```
2022-10-17 09:16:54.645 DEBUG tokio-runtime-worker sub-libp2p: [charlie] Libp2p <= Starting random Kademlia request for PeerId("1AbHqR4KoozFoA82wnYtZKyg8rg94EjsT9GSaGKATdSnZy") 
```
But  we still see this message for some validators (not all).
```  
2022-10-17 09:16:54.648 DEBUG tokio-runtime-worker sub-libp2p: [nutritious-drug-9580] Libp2p => Connected(PeerId("12D3KooWPKzmmE2uYgF3z13xjpbFTp63g9dZFag8pG6MgnpSLF4S"))    
```

**Comment by bkchr** [comment]
> But we still see this message for some validators (not all).
> 
> ```
> 2022-10-17 09:16:54.648 DEBUG tokio-runtime-worker sub-libp2p: [nutritious-drug-9580] Libp2p => Connected(PeerId("12D3KooWPKzmmE2uYgF3z13xjpbFTp63g9dZFag8pG6MgnpSLF4S"))    
> ```

This is just the normal connect message? Or why is it special? 

I briefly scanned the Kademlia code now. I think what happens is that the authority discovery is putting/getting values from the DHT. This is done through Kademlia. When we are not connected to the peers we want to push the value too, we will connect to them: https://docs.rs/libp2p-kad/0.41.0/src/libp2p_kad/behaviour.rs.html#2345 Kademlia learns about all bootnodes and reserved nodes on startup. IDK what bootnodes zombienet is setting, but I assume all validators or something? If yes, then all collators should know these nodes at least and will probably start connections to them to at leat push/get values from the DHT via Kademlia.

**Comment by mxinden** [comment]
> Maybe @tomaka / @mxinden could shed some light onto this.

Happy to help. From briefly scanning this issue I don't quite understand what the problem / goal is. Would you mind posting a summary to catch me up?

**Comment by ordian** [comment]
> > Maybe @tomaka / @mxinden could shed some light onto this.
> 
> Happy to help. From briefly scanning this issue I don't quite understand what the problem / goal is. Would you mind posting a summary to catch me up?

Hey :wave: 

In this zombienet test we're running a small relay chain with 1 full node (ferdie) and 13 validators. All validators have `--reserved-only --reserved-nodes "<ferdie>"`. There are also 3 cumulus nodes, two of them who do not produce para-blocks have the same flags on the relay-chain side of the collator. Given the flags, collators should not be connecting to any of the validators. But somehow they do at the beginning.

**Review by bkchr — APPROVED** [review]
It would be nice to understand the internals and who requests the connection. However, I get that this isn't really important here as we managed to make it fail. So, let's continue with this pr.

**Review by skunert — APPROVED** [review]

**Review by pepoviola — APPROVED** [review]

**Comment by ordian** [comment]
> `--reserved-only` and `--reserved-nodes` are only relevant when it comes to syncing/transactions/grandpa (more specifically syncing, but transactions and grandpa simply blindly follow the list of peers we use for syncing). They intentionally don't apply to Kademlia because, as you've realized, authority-discovery can't work otherwise. They also don't apply to anything higher level such as the functions used by the parachains networking.

Sure, it's fine for authority discovery to work in that case. But I didn't expect for connections to be automatically established. I'm aware of the high-level code that issues connections, e.g. collator-protocol connecting collators to the backing group. In our case, these two collators should not be issuing them because they do not produce any collation, and even if they do connect, backing group size is set to 1 at genesis and they connect to at least 5 validators.

**Review by alvicsam — APPROVED** [review]

**Review by radupopa2010 — APPROVED** [review]

**Pull request merged**